python/lsst/daf/butler/registry/interfaces/_database.py

```diff
@@ -170,6 +170,100 @@ def addInitializer(self, initializer: Callable[[Database], None]) -> None:
         self._initializers.append(initializer)
 
 
+class Session:
+    """Class representing a persistent connection to a database.
+
+    Parameters
+    ----------
+    db : `Database`
+        Database instance.
+
+    Notes
+    -----
+    Instances of Session class should not be created by client code, the
+    `Database.session` should be used to create context for a session::
+
+        with db.session() as session:
+            session.method()
+            db.method()
+
+    In current implementation sessions can be nested and transactions can be
+    nested within a session. All nested sessions and transaction share the
+    same database connection.
+
+    Session class represents a limited subset of database API that requires
+    persistent connection to a database (e.g. temporary tables which have
+    lifetime of a session). Potentially most of the database API could be
+    associated with a Session class.
+    """
+    def __init__(self, db: Database):
+        self._db = db
```

Comment:
Could the `Session` object hold its `Connection`, instead of having `Database` have an optional `Connection` that's really associated with a particular `Session`?  I bet that would avoid some `None`-checking, and it'd be a tiny step forward for thread/async safety (in that if we resolved other issues, there could be multiple parallel sessions associated with a single `Database`).

Is the problem that we would have then have to move the `transaction` interface here, too, and that's more disruptive (i.e. it's a big step towards your "solution 3")?

Reply:
I think that would require full "solution 3", basically any method that needs a connection should be moved into Session class. I agree that having Connection here is the most natural and safer way which reduces unintentional sharing but it would be a big change in API and require manual session management in all methods that did not do it before.